text/0225-configuration.md

```diff
@@ -0,0 +1,330 @@
+* The configuration format must support structured data. At the minimum arrays and associative arrays.
+* The format must support at least null, boolean, string, double precision floating point (IEEE 754-1985), and signed 64 bit integer value types.
+* Custom span processors, exporters, samplers, or other user defined extension components can be configured using this format.
+* Configure SDK, but also configure instrumentation.
+* Must offer stability guarantees while supporting evolution.
+* The structure of the configuration can be validated via a schema.
+* Support environment variable substitution to give users the option to avoid storing secrets in these files.
+
+#### Out of scope
+
+* Embedding additional configuration files within a configuration file through an expression. Additional configuration providers MAY choose to support this use-case in the future.
+
+## Internal details
+
+The schema for OpenTelemetry configuration is to be published in a repository to allow language implementations to leverage that definition to automatically generate code and/or validate end-user configuration. This will ensure that all implementations provide a consistent experience for any version of the schema they support. An example of such a proposed schema is available [here](https://github.com/MrAlias/otel-schema/tree/main/json_schema/schema).
+
+The working group proposes the use of [JSON Schema](https://json-schema.org/) as the language to define the schema. It provides:
+
+* support for client-side validation
+* code generation
+* broad support across languages
+
+In order to provide a minimal API surface area, implementations *MUST* support the following methods.
+
+### Configure(config)
+
+An API called `Configure` receives a configuration object. This method then applies the configuration object's details to the SDK. This method specifically applies the configuration object to allow for multiple configuration format providers to be supported in the future. This OTEP describes two such providers in a file and data structure formats below, but remote file formats *MAY* be implemented in the future.
+
+### ParseAndValidateConfigurationFromFile(filepath, format) -> config
+
+An API called `ParseAndValidateConfigurationFromFile` receives a string parameter indicating the file path containing the configuration to be parsed. An optional format parameter may be provided to indicate the format that this configuration uses. YAML *SHOULD* be supported, but JSON, which is a strict subset of YAML, *MAY* be supported if a language ecosystem has good reason. For example a language that has built-in support for JSON running in an environment that is sensitive to additional dependencies *MAY* only be able to support JSON configuration.
+
+The method returns a `Configuration` model that has been validated. This API *MAY* return an error or raise an exception, whichever is idiomatic to the implementation for the following reasons:
```

Comment:
What is the intended outcome? It's not clear to me from the text. I see two obvious options

1. Process starts without telemetry
2. Process crashes

Reply:
There are two cases to consider:

1. A user is calling `Configure(config)`. In this case, they can handle the exception or error in whatever way they see fit, including crashing the process or starting with some alternative fallback SDK. 
2. A user is using some sort of auto instrumentation solution like the otel java agent. Some OpenTelemetry tooling is calling `Configure(config)` and must decide how to handle error. In these cases, consider that the spec is [currently ambiguous](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/error-handling.md#basic-error-handling-principles) on whether bad user / environment config should cause the application to fail. The "MAY" implies that both failing fast and doing something else (presumably running with a noop implementation) are both acceptable. I don't think file based configuration has anything new to offer that conversation and should inherit whatever the spec says currently about failing with invalid environment variable configuration.

> The API or SDK MAY fail fast and cause the application to fail on initialization, e.g. because of a bad user config or environment, but MUST NOT cause the application to fail later at run time, e.g. due to dynamic config settings received from the Collector.